### PR TITLE
Feature/121- implement bottom Navigation, add home screen and navigation drawer

### LINKED ISSUE #121: Implement hamburger Menu

### PR DESCRIPTION
**Pull request by martin-lenweiter**
The home screen consists only of the logo and a button that goes to the learning. I thought it looked better than with also the login button (now only in the navigation drawer) but let me know if you don't like it.

### CODE REVIEW

**Comment by Nitwix** [comment]
Some tests do not pass on my phone: 
- `navigateToDisplayDatasetAndThenNavigateUpGoesToDatasetsOverview`
- `createDatasetAndNavigateUpGoesToDatasetsOverview`

The relevant error seems to be the following : `androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: with content description text: is "Navigate up"`

File: app/src/main/res/navigation/nav_graph.xml [hunk 1]
```diff
@@ -3,24 +3,12 @@
...
-            app:destination="@id/learningDatasetSelectionFragment" />
-        <action
-            android:id="@+id/action_mainFragment_to_datasetsOverviewFragment"
-            app:destination="@id/datasetsOverviewFragment" />
-    </fragment>
     <fragment
         android:id="@+id/learningDatasetSelectionFragment"
         android:name="com.github.HumanLearning2021.HumanLearningApp.view.learning.LearningDatasetSelectionFragment"
-        android:label="Select a dataset"
+        android:label="Select a dataset to start learning"
```
Comment (Nitwix) [inline app/src/main/res/navigation/nav_graph.xml]: use a string resource

File: app/src/main/res/navigation/nav_graph.xml [hunk 2]
```diff
@@ -162,8 +150,8 @@
     <fragment
         android:id="@+id/addPictureFragment"
         android:name="com.github.HumanLearning2021.HumanLearningApp.view.dataset_editing.AddPictureFragment"
-        android:label="fragment_add_picture"
-        tools:layout="@layout/fragment_add_picture" >
+        android:label="Add a new picture"
```
Comment (Nitwix) [inline app/src/main/res/navigation/nav_graph.xml]: use a string resource

File: app/src/main/res/navigation/nav_graph.xml [hunk 3]
```diff
@@ -177,4 +165,18 @@
             android:name="datasetId"
             app:argType="string" />
     </fragment>
+    <fragment
+        android:id="@+id/googleSignInWidget"
+        android:name="com.github.HumanLearning2021.HumanLearningApp.view.GoogleSignInWidget"
+        android:label="Login"
```
Comment (Nitwix) [inline app/src/main/res/navigation/nav_graph.xml]: use a string resource

File: app/src/main/res/navigation/nav_graph.xml [hunk 4]
```diff
@@ -177,4 +165,18 @@
...
+        tools:layout="@layout/fragment_google_sign_in" />
+    <fragment
+        android:id="@+id/homeFragment"
+        android:name="com.github.HumanLearning2021.HumanLearningApp.view.HomeFragment"
+        android:label="Home"
```
Comment (Nitwix) [inline app/src/main/res/navigation/nav_graph.xml]: use a string resource

**Review by Nitwix — APPROVED** [review]
Excellent changes to the UI. It feels much more usable.
I just put some minor comments

**Review by Jaakik — APPROVED** [review]
Very Intuitive design  , It just lags sometimes ( very rare ) on a Samsung A20. 

As for the tests that @Nitwix mentioned they all seem to pass when using this phone.

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/GeneralNavigationTest.kt [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                )
+            )
+    }
+
+    private fun assertCurrentFragmentIsCorrect(expected: Int) {
+        activityScenarioRule.scenario.onActivity {
+            var currentFragmentContainer =
+                it.supportFragmentManager.findFragmentById(R.id.nav_host_fragment_container)
+            val currentFragment = currentFragmentContainer?.findNavController()?.currentDestination
+            assert(currentFragment?.id == expected)
```
Comment (bbjubjub2494) [inline app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/GeneralNavigationTest.kt]: A Hamcrest `assertThat` would be nice there.

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/GeneralNavigationTest.kt [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        Intents.release()
+        activityScenarioRule.scenario.close()
+    }
+
+
+    @Test // this one used to cause trouble
+    fun navigateToDisplayDatasetAndThenNavigateUpGoesToDatasetsOverview() {
+        navigateToDisplayDataset()
+        assertCurrentFragmentIsCorrect(R.id.displayDatasetFragment)
+        onView(withContentDescription("Navigate up")).perform(click());
```
Comment (bbjubjub2494) [inline app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/GeneralNavigationTest.kt]:
```suggestion
        onView(withContentDescription("Navigate up")).perform(click())
```
semicolons are redundant in KT. This goes for all in this file

**Review by bbjubjub2494 — APPROVED** [review]
user experience is great!

optional stylistic remarks, feel free to disregard

**Comment by martin-lenweiter** [comment]
> Some tests do not pass on my phone:
> 
>     * `navigateToDisplayDatasetAndThenNavigateUpGoesToDatasetsOverview`
> 
>     * `createDatasetAndNavigateUpGoesToDatasetsOverview`
> 
> 
> The relevant error seems to be the following : `androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: with content description text: is "Navigate up"`

Is your phone in French? I will make it language independent.

File: app/src/main/res/navigation/nav_graph.xml (hunk 1 above)
Comment (martin-lenweiter) [reply app/src/main/res/navigation/nav_graph.xml]: Sorry I always forget. Thanks

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/GeneralNavigationTest.kt (hunk 6 above)
Comment (martin-lenweiter) [reply app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/GeneralNavigationTest.kt]: Yup sorry I copied some java code from stackoverflow.

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/GeneralNavigationTest.kt (hunk 5 above)
Comment (martin-lenweiter) [reply app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/GeneralNavigationTest.kt]: done thanks

**Comment by Nitwix** [comment]
> > Some tests do not pass on my phone:
> > ```
> > * `navigateToDisplayDatasetAndThenNavigateUpGoesToDatasetsOverview`
> > 
> > * `createDatasetAndNavigateUpGoesToDatasetsOverview`
> > ```
> > 
> > 
> > The relevant error seems to be the following : `androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: with content description text: is "Navigate up"`
> 
> Is your phone in French? I will make it language independent.

The tests pass once I change the language on my phone to english. No need to fix it, I'll just put my phone in english for testing, no big deal

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -58,4 +58,19 @@
...
+    <string name="nav_graph_label_learning_dataset_selection">Select a Dataset</string>
+    <string name="nav_graph_label_learning_settings">Learning Settings</string>
+    <string name="nav_graph_label_learning">Learning</string>
+    <string name="nav_graph_label_categories_editing">Edit Categories</string>
+    <string name="nav_graph_label_display_dataset">Display Datasets</string>
+    <string name="nav_graph_label_display_image_set">Dataset Images</string>
+    <string name="nav_graph_label_display_image"></string>
+    <string name="nav_graph_label_add_picture">Add a New Picture</string>
+    <string name="nav_graph_label_take_picture"></string>
+    <string name="nav_graph_label_select_picture"></string>
```
Comment (bbjubjub2494) [inline app/src/main/res/values/strings.xml]: how come those are empty?

Comment (martin-lenweiter) [reply app/src/main/res/values/strings.xml]: my mistake.

**Pull request merged**